File: trading-engine/src/trading_engine/executor.py
"""IBKR 発注エグゼキューター

Paper Trading のみ。全注文は RiskEngine を通過済みであること。
"""

import asyncio
import uuid
from datetime import datetime

from ib_insync import IB, Contract, MarketOrder, LimitOrder
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

import sys

sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parents[3] / "mcp-server" / "src"))
from claude_trade.models import Order, Decision  # noqa: E402

from trading_engine.config import settings
from trading_engine.portfolio_calc import OrderRequest

engine = create_async_engine(settings.database_url, echo=False)
async_session = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)


class Executor:
    def __init__(self):
        self._ib = IB()

    async def connect(self):
        await asyncio.wait_for(
            self._ib.connectAsync(
                host=settings.ib_host,
                port=settings.ib_port,
                clientId=settings.ib_client_id,
                readonly=False,
            ),
            timeout=10,
        )
        self._ib.reqMarketDataType(4)  # delayed-frozen data

    def disconnect(self):
        if self._ib.isConnected():
            self._ib.disconnect()

    async def execute_order(self, order_req: OrderRequest, decision_id: uuid.UUID) -> Order:
        """注文を IBKR に送信し、DB に記録する"""
        contract = Contract(
            symbol=order_req.symbol,
            secType="STK",
            exchange=order_req.exchange,
            currency=order_req.currency,
        )
        self._ib.qualifyContracts(contract)

        ib_order = MarketOrder(
            action=order_req.side,
            totalQuantity=order_req.quantity,
        )

        # DB に注文記録を作成
        db_order = Order(
            decision_id=decision_id,
            signal_id=uuid.UUID(order_req.signal_id) if order_req.signal_id else None,
            symbol=order_req.symbol,
            side=order_req.side,
            quantity=order_req.quantity,
            order_type="MKT",
            status="submitted",
        )

        try:
            trade = self._ib.placeOrder(contract, ib_order)
            db_order.ib_order_id = trade.order.orderId
            db_order.status = "submitted"

            # 約定を最大30秒待つ
            for _ in range(30):
                await asyncio.sleep(1)
                if trade.isDone():
                    break

            if trade.orderStatus.status == "Filled":
                db_order.status = "filled"
                db_order.fill_price = trade.orderStatus.avgFillPrice
                db_order.fill_quantity = int(trade.orderStatus.filled)
                db_order.filled_at = datetime.utcnow()
            elif trade.orderStatus.status == "Cancelled":
                db_order.status = "cancelled"
                db_order.error_message = trade.orderStatus.whyHeld or "Cancelled by IB"
            else:
                db_order.status = trade.orderStatus.status.lower()
        except Exception as e:
            db_order.status = "rejected"
            db_order.error_message = str(e)

        async with async_session() as session:
            session.add(db_order)
            await session.commit()
            await session.refresh(db_order)

        return db_order

    async def get_current_positions(self) -> dict[str, int]:
        """現在のポジションを {symbol: quantity} で返す"""
        portfolio = self._ib.portfolio()
        return {p.contract.symbol: int(p.position) for p in portfolio}

    async def get_nav(self) -> tuple[float, str]:
        """NAV (Net Liquidation Value) と基準通貨を取得"""
        summary = self._ib.accountSummary()
        for item in summary:
            if item.tag == "NetLiquidation" and item.currency not in ("BASE", ""):
                return float(item.value), item.currency
        return 0.0, "JPY"

    async def get_fx_rate(self, from_currency: str, to_currency: str) -> float:
        """為替レートを取得する (例: JPY→USD)"""
        if from_currency == to_currency:
            return 1.0

        # IB の Forex ペアで取得
        pair = f"{from_currency}.{to_currency}"
        contract = Contract(
            secType="CASH",
            symbol=from_currency,
            currency=to_currency,
            exchange="IDEALPRO",
        )
        try:
            self._ib.qualifyContracts(contract)
            ticker = self._ib.reqMktData(contract)
            await asyncio.sleep(3)
            self._ib.cancelMktData(contract)
            if ticker.last == ticker.last:
                return ticker.last
            elif ticker.close == ticker.close:
                return ticker.close
        except Exception:
            pass

        # フォールバック: 逆ペアで取得
        try:
            contract = Contract(
                secType="CASH",
                symbol=to_currency,
                currency=from_currency,
                exchange="IDEALPRO",
            )
            self._ib.qualifyContracts(contract)
            ticker = self._ib.reqMktData(contract)
            await asyncio.sleep(3)
            self._ib.cancelMktData(contract)
            if ticker.last == ticker.last:
                return 1.0 / ticker.last
            elif ticker.close == ticker.close:
                return 1.0 / ticker.close
        except Exception:
            pass

        return 0.0

    async def get_current_prices(
        self, symbols: list[str], exchange: str = "SMART", currency: str = "USD",
    ) -> dict[str, float]:
        """複数銘柄の現在価格を取得"""
        prices = {}
        for symbol in symbols:
            contract = Contract(symbol=symbol, secType="STK", exchange=exchange, currency=currency)
            self._ib.qualifyContracts(contract)
            ticker = self._ib.reqMktData(contract)
            await asyncio.sleep(3)
            self._ib.cancelMktData(contract)
            # NaN check: try last, then close
            if ticker.last == ticker.last:
                prices[symbol] = ticker.last
            elif ticker.close == ticker.close:
                prices[symbol] = ticker.close
        return prices

File: packages/server/src/trading-engine/executor.ts
/** IBKR 発注エグゼキューター — Paper Trading のみ */

import {
  type Contract,
  EventName,
  IBApi,
  type Order as IBOrder,
  type OrderAction,
  OrderType,
  SecType,
} from "@stoqey/ib";
import type { TickType } from "@stoqey/ib";
import type { OrderRequest } from "../broker/types.js";
import { loadEnv } from "../config.js";
import { db } from "../db/client.js";
import { type Order, orders } from "../db/schema.js";

// TickType numeric values
const TICK_BID = 1;
const TICK_ASK = 2;
const TICK_LAST = 4;
const TICK_CLOSE = 9;

let nextReqId = 2000;
function getNextReqId(): number {
  return nextReqId++;
}

export class Executor {
  private ib: IBApi;
  private connected = false;

  constructor() {
    const env = loadEnv();
    this.ib = new IBApi({
      host: env.IB_HOST,
      port: env.IB_PORT,
      clientId: 2, // different from MCP server
    });
  }

  async connect(): Promise<void> {
    if (this.connected) return;

    return new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => reject(new Error("IB connection timeout")), 10000);

      this.ib.once(EventName.connected, () => {
        clearTimeout(timeout);
        this.connected = true;
        this.ib.reqMarketDataType(1); // live (falls back to delayed if unavailable)
        resolve();
      });

      this.ib.once(EventName.error, (err: Error) => {
        clearTimeout(timeout);
        reject(err);
      });

      this.ib.connect();
    });
  }

  disconnect(): void {
    if (this.connected) {
      this.ib.disconnect();
      this.connected = false;
    }
  }

  async executeOrder(orderReq: OrderRequest, decisionId: string): Promise<Order> {
    const contract: Contract = {
      symbol: orderReq.symbol,
      secType: SecType.STK,
      exchange: "SMART",
      currency: "USD",
    };

    const ibOrder: IBOrder = {
      action: orderReq.side as OrderAction,
      totalQuantity: orderReq.quantity,
      orderType: OrderType.MKT,
    };

    let dbOrder: Order;

    try {
      // Get next valid order ID
      const orderId = await new Promise<number>((resolve) => {
        this.ib.once(EventName.nextValidId, (id: number) => resolve(id));
        this.ib.reqIds();
      });
      ibOrder.orderId = orderId;

      this.ib.placeOrder(orderId, contract, ibOrder);

      // Wait for fill (up to 30 seconds)
      const fillResult = await new Promise<{
        status: string;
        fillPrice?: number;
        fillQty?: number;
        error?: string;
      }>((resolve) => {
        const timeout = setTimeout(() => resolve({ status: "submitted" }), 30000);

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (this.ib as any).on(
          EventName.orderStatus,
          (
            id: number,
            status: string,
            filled: number,
            _remaining: number,
            avgFillPrice: number,
          ) => {
            if (id !== orderId) return;
            if (status === "Filled") {
              clearTimeout(timeout);
              resolve({
                status: "filled",
                fillPrice: avgFillPrice,
                fillQty: filled,
              });
            } else if (status === "Cancelled") {
              clearTimeout(timeout);
              resolve({ status: "cancelled", error: "Cancelled by IB" });
            }
          },
        );
      });

      [dbOrder] = await db
        .insert(orders)
        .values({
          decisionId,
          signalId: orderReq.signalId,
          symbol: orderReq.symbol,
          side: orderReq.side,
          quantity: orderReq.quantity,
          orderType: "MKT",
          status: fillResult.status,
          ibOrderId: orderId,
          fillPrice: fillResult.fillPrice ?? null,
          fillQuantity: fillResult.fillQty ?? null,
          filledAt: fillResult.status === "filled" ? new Date() : null,
          errorMessage: fillResult.error ?? null,
        })
        .returning();
    } catch (e) {
      [dbOrder] = await db
        .insert(orders)
        .values({
          decisionId,
          signalId: orderReq.signalId,
          symbol: orderReq.symbol,
          side: orderReq.side,
          quantity: orderReq.quantity,
          orderType: "MKT",
          status: "rejected",
          errorMessage: String(e),
        })
        .returning();
    }

    return dbOrder;
  }

  async getCurrentPositions(): Promise<Map<string, number>> {
    return new Promise<Map<string, number>>((resolve) => {
      const positions = new Map<string, number>();

      const handler = (_account: string, contract: Contract, pos: number) => {
        if (contract.symbol) {
          positions.set(contract.symbol, pos);
        }
      };

      const endHandler = () => {
        this.ib.removeListener(EventName.position, handler);
        this.ib.removeListener(EventName.positionEnd, endHandler);
        resolve(positions);
      };

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this.ib as any).on(EventName.position, handler);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this.ib as any).on(EventName.positionEnd, endHandler);
      this.ib.reqPositions();

      setTimeout(() => {
        this.ib.removeListener(EventName.position, handler);
        this.ib.removeListener(EventName.positionEnd, endHandler);
        this.ib.cancelPositions();
        resolve(positions);
      }, 5000);
    });
  }

  async getNav(): Promise<{ nav: number; currency: string }> {
    const reqId = getNextReqId();

    return new Promise<{ nav: number; currency: string }>((resolve) => {
      let result = { nav: 0, currency: "JPY" };

      const handler = (
        _reqId: number,
        _account: string,
        tag: string,
        value: string,
        currency: string,
      ) => {
        if (_reqId !== reqId) return;
        if (tag === "NetLiquidation" && currency !== "BASE" && currency !== "") {
          result = { nav: Number.parseFloat(value), currency };
        }
      };

      const endHandler = (_reqId: number) => {
        if (_reqId !== reqId) return;
        this.ib.removeListener(EventName.accountSummary, handler);
        this.ib.removeListener(EventName.accountSummaryEnd, endHandler);
        this.ib.cancelAccountSummary(reqId);
        resolve(result);
      };

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this.ib as any).on(EventName.accountSummary, handler);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this.ib as any).on(EventName.accountSummaryEnd, endHandler);
      this.ib.reqAccountSummary(reqId, "All", "NetLiquidation");

      setTimeout(() => {
        this.ib.removeListener(EventName.accountSummary, handler);
        this.ib.removeListener(EventName.accountSummaryEnd, endHandler);
        this.ib.cancelAccountSummary(reqId);
        resolve(result);
      }, 5000);
    });
  }

  async getFxRate(fromCurrency: string, toCurrency: string): Promise<number> {
    if (fromCurrency === toCurrency) return 1.0;

    const rate = await this._fetchFxRate(fromCurrency, toCurrency);
    if (rate > 0) return rate;

    // Fallback: reverse pair
    const reverseRate = await this._fetchFxRate(toCurrency, fromCurrency);
    if (reverseRate > 0) return 1.0 / reverseRate;

    return 0;
  }

  private async _fetchFxRate(symbol: string, currency: string): Promise<number> {
    const reqId = getNextReqId();
    const contract: Contract = {
      secType: "CASH" as SecType,
      symbol,
      currency,
      exchange: "IDEALPRO",
    };

    return new Promise<number>((resolve) => {
      let bid = 0;
      let ask = 0;
      let last = 0;

      const handler = (tickerId: number, tickType: TickType, value: number) => {
        if (tickerId !== reqId) return;
        if (isNaN(value) || value <= 0) return;
        const tt = tickType as unknown as number;
        if (tt === TICK_BID) bid = value;
        else if (tt === TICK_ASK) ask = value;
        else if (tt === TICK_LAST) last = value;
        else if (tt === TICK_CLOSE && last === 0) last = value;
      };

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (this.ib as any).on(EventName.tickPrice, handler);
      try {
        this.ib.reqMktData(reqId, contract, "", true, false);
      } catch {
        resolve(0);
        return;
      }

      setTimeout(() => {
        this.ib.cancelMktData(reqId);
        this.ib.removeListener(EventName.tickPrice, handler);
        // FX: prefer last, then bid/ask midpoint
        if (last > 0) resolve(last);
        else if (bid > 0 && ask > 0) resolve((bid + ask) / 2);
        else if (bid > 0) resolve(bid);
        else if (ask > 0) resolve(ask);
        else resolve(0);
      }, 3000);
    });
  }

  async getCurrentPrices(
    symbols: string[],
    exchange = "SMART",
    currency = "USD",
  ): Promise<Map<string, number>> {
    const prices = new Map<string, number>();

    for (const symbol of symbols) {
      const reqId = getNextReqId();
      const contract: Contract = {
        symbol,
        secType: SecType.STK,
        exchange,
        currency,
      };

      const price = await new Promise<number>((resolve) => {
        let p = 0;

        const handler = (tickerId: number, tickType: TickType, value: number) => {
          if (tickerId !== reqId) return;
          if (isNaN(value) || value <= 0) return;
          const tt = tickType as unknown as number;
          if (tt === TICK_LAST) p = value;
          else if (tt === TICK_CLOSE && p === 0) p = value;
        };

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (this.ib as any).on(EventName.tickPrice, handler);
        this.ib.reqMktData(reqId, contract, "", true, false);

        setTimeout(() => {
          this.ib.cancelMktData(reqId);
          this.ib.removeListener(EventName.tickPrice, handler);
          resolve(p);
        }, 3000);
      });

      if (price > 0) prices.set(symbol, price);
    }

    return prices;
  }
}
